### PR TITLE
Get the WDL stats for a position

### PR DESCRIPTION
**Pull request by johndoknjas**
This function allows the user to get Stockfish's win/draw/loss statistics for the current position. The WDL stats are another way of communicating the evaluation of a position. They describe the number of wins, draws, and losses that the side to move would approximately get out of 1000 games. More details on this can be found in Stockfish's source code, at the win_rate_model function of uci.cpp.

### CODE REVIEW

File: stockfish/models.py [hunk 1]
```diff
@@ -31,6 +31,7 @@ def __init__(
             "UCI_Chess960": "false",
             "UCI_LimitStrength": "false",
             "UCI_Elo": 1350,
+            "UCI_ShowWDL": "true",
```
Comment (zhelyabuzhsky) [inline stockfish/models.py:34]: Let's leave default `false`

File: stockfish/models.py [hunk 2]
```diff
@@ -283,6 +297,38 @@ def is_move_correct(self, move_value: str) -> bool:
                 else:
                     return True
 
+    def get_WDL_stats(self) -> List:
```
Comment (zhelyabuzhsky) [inline stockfish/models.py:300]: Let's keep the names of all functions in lowercase.

File: tests/stockfish/test_models.py [hunk 3]
```diff
@@ -408,3 +408,24 @@ def test_make_moves_transposition_table_speed(self):
             total_time_calculating_second += default_timer() - start
 
         assert total_time_calculating_first < total_time_calculating_second
+
+    def test_get_WDL_stats(self):
```
Comment (zhelyabuzhsky) [inline tests/stockfish/test_models.py:412]: Let's keep the names of all functions in lowercase.

File: stockfish/models.py [hunk 4]
```diff
@@ -105,6 +109,16 @@ def _go(self) -> None:
     def _go_time(self, time: int) -> None:
         self._put(f"go movetime {time}")
 
+    def _has_UCI_ShowWDL_option(self) -> bool:
```
Comment (zhelyabuzhsky) [inline stockfish/models.py:112]: Let's keep the names of all functions in lowercase.

Comment (zhelyabuzhsky) [inline stockfish/models.py:112]: Let's make this method public.

File: stockfish/models.py (hunk 1 above)
Comment (zhelyabuzhsky) [reply stockfish/models.py:34]: Let's create methods to set this property: `enable_show_wdl` and `disable_show_wdl`

File: stockfish/models.py (hunk 4 above)
Comment (zhelyabuzhsky) [reply stockfish/models.py:112]: Let's rename this method to `is_uci_whow_wdl_enabled` for example.

File: stockfish/models.py [hunk 5]
```diff
@@ -283,6 +297,38 @@ def is_move_correct(self, move_value: str) -> bool:
...
+                return []
+            elif "multipv" in current_line:
+                index_of_multipv = current_line.index("multipv")
+                if current_line[index_of_multipv + 1] == "1":
+                    index_of_wdl = current_line.index("wdl")
+                    wdl_stats = []
+                    for i in range(1, 4):
+                        wdl_stats.append(int(current_line[index_of_wdl + i]))
+                    return wdl_stats
+        raise RuntimeError("Reached the end of the get_WDL_stats function.")
```
Comment (zhelyabuzhsky) [inline stockfish/models.py:330]: Let's create test case to catch this `RuntimeError`

**Review by zhelyabuzhsky — CHANGES_REQUESTED** [review]
Thanks for your PR :)

File: stockfish/models.py (hunk 5 above)
Comment (johndoknjas) [reply stockfish/models.py]: Thanks for your feedback, I've implemented and pushed most of the changes. 
For this RuntimeError, the code shouldn't reach it (even if the user is using an older version of SF without WDL output). So in order to make it be thrown, either there would have to be a bug in this function, or maybe the user tries doing something like multithreading. 
I also changed some of the WDL function names to be clearer for what they do. The main thing is trying to indicate that the WDL option can be enabled/disabled (for more recent SFs) or not a UCI option at all (SF 11 and older).

File: README.md [hunk 6]
```diff
@@ -104,6 +104,27 @@ stockfish.get_top_moves(3)
 ]
 ```
 
+### Get Stockfish's win/draw/loss stats for the side to move in the current position.
+```python
+stockfish.get_WDL_stats()
```
Comment (zhelyabuzhsky) [inline README.md:109]:
```suggestion
stockfish.get_wdl_stats()
```

File: stockfish/models.py [hunk 7]
```diff
@@ -42,6 +43,12 @@ def __init__(
 
         self._put("uci")
 
+        self.has_wdl_option = None  # Necessary to have before call below.
```
Comment (zhelyabuzhsky) [inline stockfish/models.py:46]: Why `None`?

Is it possible to rewrite without this line?

File: stockfish/models.py [hunk 8]
```diff
@@ -283,6 +290,78 @@ def is_move_correct(self, move_value: str) -> bool:
...
+                    index_of_wdl = current_line.index("wdl")
+                    wdl_stats = []
+                    for i in range(1, 4):
+                        wdl_stats.append(int(current_line[index_of_wdl + i]))
+                    if was_wdl_option_false_before:
+                        self.set_show_wdl_option(False)
+                    return wdl_stats
+        raise RuntimeError("Reached the end of the get_wdl_stats function.")
+
+    def does_sf_version_have_wdl_option(self) -> bool:
```
Comment (zhelyabuzhsky) [inline stockfish/models.py:332]: Add docstring please

File: stockfish/models.py [hunk 9]
```diff
@@ -42,6 +43,12 @@ def __init__(
 
         self._put("uci")
 
+        self.has_wdl_option = None  # Necessary to have before call below.
+        self.has_wdl_option = self.does_sf_version_have_wdl_option()
```
Comment (zhelyabuzhsky) [inline stockfish/models.py:47]:
```suggestion
        self._has_wdl_option = self.does_sf_version_have_wdl_option()
```

File: stockfish/models.py (hunk 8 above)
Comment (zhelyabuzhsky) [inline stockfish/models.py:332]:
```suggestion
    def has_thi_version_wdl_option(self) -> bool:
```

File: tests/stockfish/test_models.py [hunk 10]
```diff
@@ -408,3 +408,55 @@ def test_make_moves_transposition_table_speed(self):
             total_time_calculating_second += default_timer() - start
 
         assert total_time_calculating_first < total_time_calculating_second
+
+    def test_get_wdl_stats(self):
+        stockfish = Stockfish(depth=15, parameters={"MultiPV": 2})
+        if stockfish.does_sf_version_have_wdl_option():
```
Comment (zhelyabuzhsky) [inline tests/stockfish/test_models.py:414]:
```suggestion
    def test_get_wdl_stats(self, stockfish):
        if stockfish.does_sf_version_have_wdl_option():
```

File: tests/stockfish/test_models.py [hunk 11]
```diff
@@ -408,3 +408,55 @@ def test_make_moves_transposition_table_speed(self):
...
+            assert wdl_stats[0] > wdl_stats[2] * 1.8
+            assert stockfish._parameters["UCI_ShowWDL"] == "false"
+            stockfish.set_fen_position("8/8/8/8/8/6k1/6p1/6K1 w - - 0 1")
+            assert stockfish.get_wdl_stats() is None
+        else:
+            with pytest.raises(RuntimeError):
+                stockfish.get_wdl_stats()
+
+    def test_does_sf_version_have_wdl_option(self):
+        stockfish = Stockfish()
```
Comment (zhelyabuzhsky) [inline tests/stockfish/test_models.py:436]:
```suggestion
    def test_does_sf_version_have_wdl_option(self, stockfish):
```

File: tests/stockfish/test_models.py [hunk 12]
```diff
@@ -408,3 +408,55 @@ def test_make_moves_transposition_table_speed(self):
...
+        if stockfish.get_stockfish_major_version() <= 11:
+            assert not stockfish.does_sf_version_have_wdl_option()
+            assert "UCI_ShowWDL" not in stockfish._parameters
+            with pytest.raises(RuntimeError):
+                stockfish.get_wdl_stats()
+
+    def test_set_show_wdl_option(self):
+        stockfish = Stockfish()
```
Comment (zhelyabuzhsky) [inline tests/stockfish/test_models.py:444]:
```suggestion
    def test_set_show_wdl_option(self, stockfish):
```

**Comment by zhelyabuzhsky** [comment]
Thank you :) Well done 👍🏻

**Review by zhelyabuzhsky — APPROVED** [review]

**Pull request merged**